File: JAVA/linked_list.java
/*

Methods implemented:

    public  T       head()
    public  void    addBeginning(T value)
    public  void    addEnd(T value)
    public  void    deleteBeginning()
    public  void    deleteEnd()
    public  int     count()
    public  void    reverse()


*/


package JAVA;

class Node<T>                   	   
{
    private T data;
    private Node<T> next;

    Node()
    {
        data = null;
        next = null;
    }
    public T getData()
    {
        return this.data;
    }
    public void setData(T da)
    {
        this.data = da;
    }
    public Node<T> getNext()
    {
        return this.next;
    }
    public void setNext(Node<T> n)
    {
        this.next = n;
    }
}

public class linked_list<T>             
{
    private Node<T> head;

    linked_list()
    {
        head = null;
    } 

    public T head()                                 // Returns the data stored in the first element
    {
        return this.head.getData();
    }

    public void addBeginning(T value)               // Add a node to beginning of the linked list
    {
        Node<T> newnode = new Node<T>();
        newnode.setData(value); 
        if (this.head != null)
        {
            newnode.setNext(this.head);
        }
        this.head = newnode;
    }

    public void addEnd(T value)                     // Add a node to end of the linked list
    {
        Node<T> newnode = new Node<T>();
        newnode.setData(value);
        if(this.head == null)
        {
            this.head = newnode;
        }
        else
        {
            Node<T> temp = new Node<T>();
            temp = this.head;
            while(temp.getNext() != null)
            {
                temp = temp.getNext();
            }
            temp.setNext(newnode);
        }
    }

    public void deleteBeginning()                   // Delete the node at beginning of the linked list
    {
        if(this.head == null)
        {
            System.err.println("Cannot delete: list already empty");
            System.exit(0);
        }
        else
        {
            Node<T> temp = new Node<T>();
            temp = this.head;
            this.head = this.head.getNext();
            temp.setData(null);
            temp.setNext(null);
            temp = null;
        }
    }

    public void deleteEnd()                         // Delete a node at end of he linked list
    {
        if (this.head == null)
        {
            System.err.println("Cannot delete: list already empty");
            System.exit(0);
        }
        else
        {
            if(this.head.getNext() == null)
            {
                this.head.setData(null);
                this.head = null;
            }
            else
            {
                Node<T> temp = new Node<T>();
                temp = this.head;
                while(temp.getNext().getNext() != null)
                {
                    temp = temp.getNext();
                }
                temp.getNext().setData(null);
                temp.setNext(null);
            }  
        }
    }

    public int count()                              // Count the number of nodes in linked list
    {
        int count = 0;
        Node<T> temp = new Node<T>();
        temp = this.head;
        while(temp != null)
        {
            count++;
            temp = temp.getNext();
        }
        return count;
    }

    public void reverse()                           // Reverse the linked list
    {
        if (this.count() == 0)
        {
            System.err.println("Error: Cannot reverse empty list");
            System.exit(0);
        }
        else if(this.count() == 1)
        {
            return;
        }
        else if(this.count() == 2)
        {
            this.head.getNext().setNext(this.head);
            this.head = this.head.getNext();
            this.head.getNext().setNext(null);
        }
        else 
        {
            Node<T> pointer1 = new Node<T>();
            Node<T> pointer2 = new Node<T>();
            Node<T> pointer3 = new Node<T>();
            pointer1 = this.head;
            pointer2 = this.head.getNext();
            pointer3 = pointer2.getNext();
            pointer1.setNext(null);
            while(pointer3 != null)
            {
                pointer2.setNext(pointer1);
                pointer1 = pointer2;
                pointer2 = pointer3;
                pointer3 = pointer3.getNext();
            }
            pointer2.setNext(pointer1);
            this.head = pointer2;
        }
    }

    public String toString()            
    {
        Node<T> temp = new Node<T>();
        temp = this.head;
        if(temp == null)
        {
            return "Empty List";
        }
        String s = "";
        while(temp != null)
        {
            s = s + temp.getData() + " ";
            temp = temp.getNext();
        }
        return s;
    }
}

File: go/linkedlist.go
package main

import "fmt"

type Node struct {
	value int
	next  *Node
}

type LinkedList struct {
	head   *Node
	length int
}

func insertHead(list *LinkedList, data int) {
	var newNode = Node{data, nil}
	if list.head == nil {
		list.head = &newNode
	} else {
		newNode.next = list.head
		list.head = &newNode
	}
	list.length++
}

func removeHead(list *LinkedList) {
	if list.head != nil {
		list.head = list.head.next
		list.length--
	}
}

func insertEnd(list *LinkedList, data int) {
	var newNode = Node{data, nil}
	if list.head == nil {
		list.head = &newNode
	}
	var tempNode *Node = list.head
	for tempNode.next != nil {
		tempNode = tempNode.next
	}
	tempNode.next = &newNode
	list.length++
}

func removeEnd(list *LinkedList) {
	var tempNode *Node = list.head
	if list.head == nil || list.head.next == nil {
		list.head = nil
		return
	}
	for tempNode.next.next != nil {
		tempNode = tempNode.next
	}
	tempNode.next = nil
	list.length--
}

func getAt(list *LinkedList, position int) int {
	if position < 0 || position >= list.length {
		return -1
	}
	var tempNode *Node = list.head
	for i := 0; i < position; i++ {
		tempNode = tempNode.next
	}
	return tempNode.value
}

func insertAt(list *LinkedList, data int, position int) {
	if position < 0 {
		return
	}
	var newNode = Node{data, nil}
	var tempNode *Node = list.head
	position--
	for position > 0 && tempNode.next != nil {
		tempNode = tempNode.next
		position--
	}
	newNode.next = tempNode.next
	tempNode.next = &newNode
}

func deleteAt(list *LinkedList, position int) {
	if position < 0 || position >= list.length {
		return
	}
	var tempNode *Node = list.head
	position--
	for position > 0 && tempNode.next != nil {
		tempNode = tempNode.next
		position--
	}
	tempNode.next = tempNode.next.next
}

func reverse(list *LinkedList) {
	if list.head == nil || list.head.next == nil {
		return
	} else if list.head.next.next == nil {
		list.head.next.next = list.head
		list.head = list.head.next
		list.head.next.next = nil
	} else {
		var ptr1 *Node = list.head
		var ptr2 *Node = ptr1.next
		var ptr3 *Node = ptr2.next
		ptr1.next = nil
		for ptr3 != nil {
			ptr2.next = ptr1
			ptr1 = ptr2
			ptr2 = ptr3
			ptr3 = ptr3.next
		}
		ptr2.next = ptr1
		list.head = ptr2
	}
}

func print(list *LinkedList) {
	var tempNode *Node = list.head
	for tempNode != nil {
		fmt.Print(tempNode.value)
		tempNode = tempNode.next
	}
	fmt.Println()
}
